placement/cluster_mapping_provider.go

```diff
@@ -0,0 +1,476 @@
+	// NB(mmihic): We use copy-on-write for the database mappings, so it's safe to return
+	// a reference without the
+	mp.RLock()
+	defer mp.RUnlock()
+
+	if len(mp.byRetention) == 0 {
+		return nil
+	}
+
+	for _, dbm := range mp.byRetention {
+		// TODO(mmihic): Handle read cutover time and cutover complete time
+		if retentionPeriodInSecs <= dbm.maxRetentionInSecs {
+			return dbm
+		}
+	}
+
+	// If we the retentionPeriod falls outside all database retention periods,
+	// use the one with the longest period
+	return mp.byRetention[len(mp.byRetention)-1]
+}
+
+// update updates the mappings based on new Placement information
+func (mp *clusterMappingProvider) update(p *schema.Placement) error {
+	// Avoid the need for upstream code to synchronize on the cluster
+	// mappings by first making a copy, then updating the copy in place,
+	// then swapping out the pointers atomically
+	mp.RLock()
+	byName := make(map[string]*databaseMappings, len(mp.byName))
+	byRetention := make([]*databaseMappings, len(mp.byRetention))
+	for n, dbm := range mp.byRetention {
+		dbm := dbm.clone()
+		byRetention[n], byName[dbm.name] = dbm, dbm
+	}
+	mp.RUnlock()
+
+	// Update the clone with the new settings
+	for _, db := range p.Databases {
+		if existing := byName[db.Properties.Name]; existing != nil {
+			if err := existing.update(db); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// This is a new mapping - create and apply all rules
+		dbm, err := newDatabaseMappings(db, mp.log)
+		if err != nil {
+			return err
+		}
+
+		byName[db.Properties.Name] = dbm
+		byRetention = append(byRetention, dbm)
+	}
+
+	sort.Sort(databaseMappingsByMaxRetention(byRetention))
+
+	// Swap out the pointers
+	mp.Lock()
+	mp.byRetention, mp.byName = byRetention, byName
+	mp.Unlock()
+	return nil
+}
+
+// watchPlacementChanges is a background goroutine that watches
+// for placement changes
+func (mp *clusterMappingProvider) watchPlacementChanges() {
+	var placement schema.Placement
+	for range mp.watch.C() {
+		val := mp.watch.Get()
+		if err := val.Unmarshal(&placement); err != nil {
+			mp.log.Errorf("could not unmarshal placement data: %v", err)
+			continue
+		}
+
+		mp.log.Infof("received placement version %d", val.Version())
+		mp.update(&placement)
+	}
+	close(mp.closed)
+}
+
+// databaseMappings are the mappings for a given database.
+type databaseMappings struct {
+	name               string
+	version            int32
+	maxRetentionInSecs int32
+	active             []*activeClusterMapping
+	log                xlog.Logger
+}
+
+// newDatabaseMappings returns a newly initialized set of databaseMappings
+func newDatabaseMappings(db *schema.Database, log xlog.Logger) (*databaseMappings, error) {
+	dbm := &databaseMappings{
+		name:               db.Properties.Name,
+		version:            db.Version,
+		maxRetentionInSecs: db.Properties.MaxRetentionInSecs,
+		log:                log,
+	}
+
+	for _, rules := range db.MappingRules {
```

Comment:
The are guaranteed sorted by the layer that writes them, but sure we can sort them here as well.